### ISSUE TITLE
add Action Dropout on minerva WN18RR

### ISSUE THREAD

**Issue by David-Lee-1990** [salesforce/MultiHopKG#12]
Hi, I tried action dropout trick on original minerva code on WN18RR. However,  hit@10 decreased from 0.52 to 0.38 when action dropout rate changed from 1.0 to 0.9. Is there any other auxiliary tricks for action dropout?

**Comment by todpole3** [comment]
You mean changing action dropout rate from 0.0 to 0.1? 0.9 is very aggressive dropout and 1.0 implies dropout everything and randomly sample an edge.

**Comment by David-Lee-1990** [comment]
@todpole3  I  re-edited my issue to show more detailed information of the training results.

**Comment by todpole3** [comment]
@David-Lee-1990 In MINERVA code, is `pre_distribution` used in gradient computation?

For us we only use action dropout to encourage diverse sampling but the policy gradient is still computed using the original probability vector.

**Comment by David-Lee-1990** [comment]
In MINERVA, there is no dropout and I try to add this advancement on it. Follow your idea, I use dropout to encourage diverse sampling and  the policy gradient is still compuated using the original distribution.  I  tested this for two versions:  one is relation-only and the other not. Both versions show the similar results as i stated in the issue.

**Comment by todpole3** [comment]
@David-Lee-1990 My question is, after adding "action dropout", did you use the updated probability vector `pre_distribution` to compute the policy gradient?

**Comment by David-Lee-1990** [comment]
No, i use the original one.

**Comment by todpole3** [comment]
@David-Lee-1990 I cannot spot anything wrong with the code snippet you posted. Thanks for sharing. It might have something to do with the integration with the rest of MINERVA code.

Technically you only disturbed the sampling prob by a small factor (and your policy gradient computation still follows the traditional formula) so the result shouldn't change so significantly no matter what.

Would you mind run a sanity-checking experiments by setting the dropout rate to 0.01 and see how the result turned out? Technically the change should be very small.

**Comment by David-Lee-1990** [comment]
hi, i run a sanity-checking experiments  by setting the keep rate in [1.0, 0.99, 0.97, 0.95, 0.93, 0.90]. The results of hits@1 on training batch is as follows:

**1.0 VS 0.99**
![result-99](https://user-images.githubusercontent.com/29653610/63331100-8ca23800-c367-11e9-8368-ecb3b6aaacc4.png)
**1.0 VS 0.97**
![result-97](https://user-images.githubusercontent.com/29653610/63331154-a6dc1600-c367-11e9-90c5-a79e871d9e2e.png)
**1.0 VS 0.95**
![result-95](https://user-images.githubusercontent.com/29653610/63331191-b5c2c880-c367-11e9-8786-d8e2872587d6.png)
**1.0 VS 0.93**
![result-93](https://user-images.githubusercontent.com/29653610/63331208-beb39a00-c367-11e9-9748-4bf35124b9d5.png)
**1.0 VS 0.90**
![result-9](https://user-images.githubusercontent.com/29653610/63331254-dee35900-c367-11e9-8405-239ccd4c55fd.png)

**Comment by todpole3** [comment]
@David-Lee-1990 Very interesting. I want to look deeper into this issue.

The most noticeable difference is that the dev result you reported without action dropout is close to what we have with 0.1 action dropout and significantly higher than what we have without action dropout.

Besides action dropout rate, did you use the same set of hyperparameters as we did in the configuration files?
If not, would you mind sharing your set of hyperparameters? I want to see if I can reproduce the same results on our code repo.

And one more question, did you observe similar trend on other datasets using MINERVA code + action dropout?

**Comment by todpole3** [comment]
> I tested this for two versions: one is relation-only and the other not. Both versions show the similar results as i stated in the issue.

Are the plots shown above generated with `relation-only` or not?

**Comment by David-Lee-1990** [comment]
> The most noticeable difference is that the dev result you reported without action dropout is close to what we have with 0.1 action dropout and significantly higher than what we have without action dropout.

@todpole3  about the dev result, i need to clarify that i used "sum" method, which is   different with "max" method as you used, when calculating hit@k and MRR.  "sum" method  ranks the predicted entity by adding those probalibily, which predict the same end entity, up. The following code is from MINERVA.  And I also test "max" method on WN18RR, hit@k and MRR is not less than your model with ConvE. 

![image](https://user-images.githubusercontent.com/29653610/63394535-b48aad80-c3f2-11e9-954c-7f576783d897.png)

**Comment by David-Lee-1990** [comment]
> > I tested this for two versions: one is relation-only and the other not. Both versions show the similar results as i stated in the issue.
> 
> @David-Lee-1990 Are the plots shown above generated with `relation-only` or not?

relation only

**Comment by David-Lee-1990** [comment]
> Besides action dropout rate, did you use the same set of hyperparameters as we did in the configuration files?
If not, would you mind sharing your set of hyperparameters? I want to see if I can reproduce the same results on our code repo.

I give my hyperparameters in the form of your notation as follows:

group_examples_by_query="False"
use_action_space_bucketing="False"
bandwidth=200
entity_dim=100
relation_dim=100
history_dim=100
history_num_layers=1
train_num_rollouts=20
dev_num_rollouts=40
num_epochs=1000  # follow minerva, i randomly choose training data which has batch_size samples 
train_batch_size=128
dev_batch_size=128
learning_rate=0.001
grad_norm=5
emb_dropout_rate=0
ff_dropout_rate=0
action_dropout_rate=1.0
beta=0.05
relation_only="True"
beam_size=100

**Comment by David-Lee-1990** [comment]
> And one more question, did you observe similar trend on other datasets using MINERVA code + action dropout?
@todpole3  Follow your advice, I test action dropout on nell-995 today, its performance is as follows. 
![result](https://user-images.githubusercontent.com/29653610/63403696-deec6300-c412-11e9-81ca-ad792e41f640.png)

![result-mrr](https://user-images.githubusercontent.com/29653610/63404088-70100980-c414-11e9-985a-bc38920ce4ec.png)
